File: infra/src/test/java/com/beat/infra/external/storage/s3/S3FileStorageAdapterTest.java
package com.beat.infra.external.storage.s3;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.mockito.Mockito.when;
import static org.mockito.Mockito.verifyNoInteractions;

import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.model.AmazonS3Exception;
import com.amazonaws.services.s3.model.ObjectMetadata;
import com.beat.contracts.storage.ImageObjectMetadata;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.test.util.ReflectionTestUtils;

@ExtendWith(MockitoExtension.class)
class S3FileStorageAdapterTest {

	@Mock
	private AmazonS3 amazonS3;

	@Test
	void findImageObjectMetadataReturnsHeadObjectMetadata() {
		S3FileStorageAdapter adapter = adapter();
		ObjectMetadata metadata = new ObjectMetadata();
		metadata.setContentType("image/png");
		metadata.setContentLength(1024L);
		when(amazonS3.getObjectMetadata("bucket", "dev/carousel/image.png")).thenReturn(metadata);

		ImageObjectMetadata result = adapter.findImageObjectMetadata("dev/carousel/image.png");

		assertEquals("image/png", result.getContentType());
		assertEquals(1024L, result.getContentLength());
	}

	@Test
	void findImageObjectMetadataReturnsNullWhenObjectDoesNotExist() {
		S3FileStorageAdapter adapter = adapter();
		AmazonS3Exception exception = new AmazonS3Exception("not found");
		exception.setStatusCode(404);
		when(amazonS3.getObjectMetadata("bucket", "dev/carousel/missing.png")).thenThrow(exception);

		assertNull(adapter.findImageObjectMetadata("dev/carousel/missing.png"));
	}

	@Test
	void findImageObjectMetadataRejectsOtherEnvironmentsWithoutHeadObject() {
		S3FileStorageAdapter adapter = adapter();

		assertNull(adapter.findImageObjectMetadata("prod/poster/poster.png"));

		verifyNoInteractions(amazonS3);
	}

	private S3FileStorageAdapter adapter() {
		S3FileStorageAdapter adapter = new S3FileStorageAdapter(amazonS3);
		ReflectionTestUtils.setField(adapter, "bucket", "bucket");
		ReflectionTestUtils.setField(adapter, "keyPrefix", "dev");
		return adapter;
	}
}

File: core/infra/src/test/kotlin/com/beat/infra/external/storage/s3/S3FileStorageAdapterTest.kt
package com.beat.infra.external.storage.s3

import com.amazonaws.services.s3.AmazonS3
import com.amazonaws.services.s3.model.AmazonS3Exception
import com.amazonaws.services.s3.model.ObjectMetadata
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import org.mockito.Mockito
import org.springframework.test.util.ReflectionTestUtils

class S3FileStorageAdapterTest : FunSpec({
    test("Promotion image object가 현재 environment에 존재하면 true를 반환한다") {
        val amazonS3 = Mockito.mock(AmazonS3::class.java)
        val adapter = adapter(amazonS3)
        val metadata = ObjectMetadata().apply {
            contentType = "image/png"
            contentLength = 1024L
        }
        Mockito.`when`(amazonS3.getObjectMetadata("bucket", "dev/carousel/image.png")).thenReturn(metadata)

        adapter.exists("dev/carousel/image.png") shouldBe true
    }

    test("S3가 404를 반환하면 Promotion image가 존재하지 않는 것으로 해석한다") {
        val amazonS3 = Mockito.mock(AmazonS3::class.java)
        val adapter = adapter(amazonS3)
        val exception = AmazonS3Exception("not found").apply { statusCode = 404 }
        Mockito.`when`(amazonS3.getObjectMetadata("bucket", "dev/carousel/missing.png")).thenThrow(exception)

        adapter.exists("dev/carousel/missing.png") shouldBe false
    }

    test("다른 environment key는 S3 조회 없이 거부한다") {
        val amazonS3 = Mockito.mock(AmazonS3::class.java)
        val adapter = adapter(amazonS3)

        adapter.exists("prod/poster/poster.png") shouldBe false
        Mockito.verifyNoInteractions(amazonS3)
    }
})

private fun adapter(amazonS3: AmazonS3): S3FileStorageAdapter =
    S3FileStorageAdapter(amazonS3).also {
        ReflectionTestUtils.setField(it, "bucket", "bucket")
        ReflectionTestUtils.setField(it, "keyPrefix", "dev")
    }
